Sources/Basics/NativePathExtensions.swift

```diff
@@ -0,0 +1,32 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import struct Foundation.URL
+import struct TSCBasic.AbsolutePath
+
+extension String {
+    public func nativePathString(escaped: Bool) -> Self {
```

Comment:
where is this used? is this windows only?

Reply:
It is mostly used in tests, but there will be some paths that need to native path convert (e.g. VFS).

Reply:
how / where are we planning to use this in PackageLoading?

Reply:
I also wonder if this should be a function of `FileSystem` rather than of path since it is by definition `fileSystemRepresentation`

Reply:
> how / where are we planning to use this in PackageLoading?

We need to normalise the paths before emitting them into the VFS.

> I also wonder if this should be a function of FileSystem rather than of path since it is by definition fileSystemRepresentation

Is there a `FileSystem` module?  But, as methods on `FileSystem` I feel would be bad ... unless we want to move `AbsolutePath` into `FileSystem` too - that is part of the `FileSystem` by definition as well.

Reply:
> can you elaborate on this?

https://github.com/apple/swift-package-manager/blob/main/Sources/PackageLoading/ManifestLoader.swift#L526-L530

The VFS doesn't have issues on Windows, SPM has issues on Windows because it doesn't understand paths.  A path is spelt with the arc separator (`\`), not a random separator (`/`).  This means that the VFS will fail to find the source files because the path is incorrect.  This has a small issue still to deal with as we also drop the drive letter which causes problems.  Really, its just that the overlay is being written incorrectly by SPM due to the path handling.

> I was imagining something like FileSystem::nativeRepresentation(AbsolutePath) and then each concrete FS can implement this logic that make sense for it. for example, some abstract FSs wont care while the "real" one would

I don't think that this is the right abstraction.  The fundamental issue is that we do not faithfully represent paths - we write them in Unix form which is a problem.  This is a representation of the `AbsolutePath` or `RelativePath`, not an intrinsic property of the `FileSystem`.  e.g. I could be using ext3 on Windows, but doesn't mean that I would use `/` as the arc separator.  This is how you represent a path in a string form vs binary form.  IIRC, we also did the same thing in `System.Path`.

Reply:
but it does not seems to me like we are actually solving anything with this additional API. The API in question is "fileSystemRepresentation" which means the representation for a concrete file system

If we have an issue about how we represent paths, and the fact it is *nix oriented, that may be worth solving and not working around?

Reply:
> If we have an issue about how we represent paths, and the fact it is *nix oriented, that may be worth solving and not working around?

This could be interesting!  That would definitely avoid this.  What did you have in mind?

This has nothing to do with a concrete file system even though it says `fileSystemRepresentation` in the name.  The reason it says that is because it has to do with the fact that you are converting a URI to a string representation of a file path.  The operation is entirely devoid of the filesystem and is a syntactic operation over the string reprensetation of a URI to a form that can be used with APIs that expect a file path.

Reply:
I would guess AbsolutePath should have in-memory representation of there path segments rather than a strigified version of them?

Reply:
Yeah, that would be good, except at that point you are effectively redesigning System.Path.

Reply:
`System.FilePath` allows both absolute and relative paths IIRC. While with TSC we have a type-level separation between those, which I guess we'd lose if we were to switch to `System.FilePath` everywhere?

Reply:
Right, and the representation issues exist in both technically, it's just that we often convert to AbsolutePath so as to avoid the needing the third instance of this.

Reply:
> Yeah, that would be good, except at that point you are effectively redesigning System.Path.

or we rebase AbsolutePath and RelativePath on top of System.Path which we wanted to do anyways

Reply:
IMO if we think absolute vs relative is meaningful enough, we should push for it in swift-system. Having a wrapper around it long term seems like a non-goal to me.

Reply:
@neonichu that seems like the right thing to me.  However, the question still stands of what to do right now.  This is needed in a couple more places, and centralising the logic seems like the right thing to me.

Reply:
agree with both of you. imo rebasing on top of path is a step towards migrating off these types or pushing them down to swift-system if we feel they are valuable

Reply:
That is a much larger undertaking and I'd like to get things into a better state before going down that road.  There's a change that's been sitting for ages because it's non-trivial.

Reply:
IMO, this PR seems fine as long as we have an idea on longer term strategy which it seems like we do.

Reply:
@compnerd would you mind creating a GH issue to track the longer term solution? also if this is not truly the a fileSystemRepresentation, what is a better name for this API?, and if it should not be part of the public API should we underscore it and add a comment that it is for internal use only and that it is likely to be deprecated soon?

Reply:
I really like the idea of the underscoring.  I'll do that and file the issue (I don't know if there is one already filed for it).

Hmm, well as you know, naming is one of the two most difficult problems in CS (along with cache invalidation and off-by-ones 😆).  I would say `OSPath` is a good name for this.  However, I am loathe to rename this to `OSPathString` because the same name is now used across multiple repositories.

Reply:
https://github.com/apple/swift-package-manager/issues/6499